Cmsis rtos v2

rgundi opened the pull request:
These patchsets introduce the CMSIS RTOS v2 API support for Zephyr.

rgundi committed:
CMSIS RTOS V2: Introduce CMSIS RTOS V2 API header file

CMSIS RTOS API provides a generic RTOS interface for embedded
processors (actually for Cortex-M processors but are generic
enough to be used elsewhere). This header file is for V2 version.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v2: Implement support for thread APIs

These APIs allow defining, creating and controlling of thread
related functionalities.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v2: Implement support for generic wait APIs

APIs to introduce wait i.e osDelay and osDelayUntil are defined
here. They are analogous to k_sleep in the kernel.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v2: Introduce tests to make use of Thread APIs

Some test programs to illustrate the usage of thread APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v2: Implement support for timer APIs

These APIs provide the support of virtual timers. All timers
can be started, restarted, or stopped. Timers can be configured
as one-shot or periodic.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v2: Introduce tests to make use of Timer APIs

Some test programs to illustrate the usage of timer APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v2: Implement support for mutex APIs

These APIs allow creating, acquiring, releasing and deleting
the mutexes.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v2: Introduce tests to make use of Mutex APIs

Some test programs to illustrate the usage of Mutex APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v2: Implement support for semaphore APIs

These APIs allow creating, acquiring, releasing and deleting
of semaphores.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v2: Introduce tests to make use of Semaphore APIs

Some test programs to illustrate the usage of Semaphore APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v2: Implement support for mempool APIs

These APIs allow creating, allocating and freeing of mempools.

Note: "Mempool" in CMSIS actually means memslabs in Zephyr.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v2: Introduce tests to make use of mempool APIs

Some test programs to illustrate the usage of mempool APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v2: Implement support for Message Queue APIs

These APIs provide message queue functionality like create,
put and get.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v2: Introduce tests to make use of messageQ APIs

Some test programs to illustrate the usage of msgq APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v2: Implement support for Event Flag APIs

Events are used to trigger execution states between threads.
These APIs provide functionalities like event set, clear and
wait.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v2: Introduce tests to make use of EventFlag APIs

Some test programs to illustrate the usage of EventFlag APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v2: Implement support for Thread Flags APIs

Thread Flags are used to trigger execution states between threads.
These APIs provide functionalities like set, clear and wait.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v2: Introduce tests to make use of ThreadFlags APIs

Some test programs to illustrate the usage of ThreadFlags APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v2: Implement support for Kernel APIs

Implement support for Kernel management APIs like
osKernelInitialize, osKernelGetTickCount, osKernelGetSysTimerCount
etc.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi commented:
Shippable is failing for 2 reasons.
1) qemu_cortex_m3 - This has multiqueue scheduler enabled and multi-queue scheduler supports only a max of 32 priority levels. Cmsis_rtos_v2 has 56 priority levels. As discussed earlier with Anas, this failure can be ignored.
2) checkpatch issues - This is because of the external header file cmsis_os2.h. This can be ignored as well.

galak reviewed (dismissed):
We need a better category for this from where we are putting the files (need to fix cmsis_v1 as well).  This would impact both the include/ and lib/.

maybe "compat_api" ?

```
include/compat_api/cmsis_v1
include/compat_api/cmsis_v2
include/compat_api/posix

lib/cmsis_rtos_v1 -> compat_api/cmsis_v1
```

Open to better names/suggestions

rgundi commented:
> > 1. As discussed earlier with Anas, this failure can be ignored.
> 
> we cant ignore tests failing, this will fail for everyone after this change is merged, so please find a solution for this issue.

Well.. not sure why this will affect others. It will only affect others if they enable CMSIS RTOS v2. The fix cannot be in CMSIS as that is not desirable and will violate the basic premise of portability. The multi-queue scheduler scheduler can be doubled (to 64) which will solve this issue but i would rather let @andyross to comment on that.

rgundi commented:
@galak.. open to suggestions. I have no qualms introducing another folder. Guess that blends in.

nashif commented:
> Well.. not sure why this will affect others.

because this is a test that fails and will not pass CI.

nashif commented:
> maybe "compat_api" ?

not compat_api please, this will be confusing, it is more about portability than compatibility (i.e. with old APIs for example)

maybe just "portable" or "portability"?

galak commented:
> > maybe "compat_api" ?
> 
> not compat_api please, this will be confusing, it is more about portability than compatibility (i.e. with old APIs for example)
> 
> maybe just "portable" or "portability"?

@pfalcon usually has good ideas for such things.  @pfalcon ideas/thoughts?

pfalcon commented:
Thanks. Actually, the only thing which dances in my mind regarding this is "Why Intel works on CMSIS APIs?" ;-).

> include/portable/cmsis_v1
>include/portability/cmsis_v1

IMHO, these look worse than:

> include/compat_api/cmsis_v1

The only thing which I can quickly think of is:

> include/emu_api/cmsis_v1

But that can be confusing too. But IMHO, having "api" part there is really useful. Perhaps @nashif can elaborate what exactly he doesn't like with "compat_api"?


Also have comment about:

> lib/cmsis_rtos_v1 -> compat_api/cmsis_v1

IMHO, currently the top level of Zephyr tree is pretty clean. If we push too much stuff to the top level, it may become less so. `lib/` is not overloaded currently, and IMHO good enough place for this.

pfalcon commented:
Oh, and let's not forget that POSIX is also "compatibility/portability API" in a sense, so any ideas re: CMSIS should also be tested for applicability to existing POSIX stuff.

dbkinder commented on doc/porting/cmsis_rtos_v2.rst:
```diff
@@ -0,0 +1,63 @@
+making it generic. For more information on CMSIS RTOS v2, please refer
+http://www.keil.com/pack/doc/CMSIS/RTOS2/html/index.html
+
+Features not supported in Zephyr implementation:
```
Needs some formatting fixes (I've laid this out as a "dictionary list", where the first line will be bold and the subsequent indented lines will be an indented paragraph):

    Features not supported in Zephyr implementation
    ***********************************************

    Kernel
       osKernelGetState, osKernelSuspend, osKernelResume are not supported.
       Also, osKernelInitialize and osKernelStart are pretty much empty.

    Threads
       osThreadJoin and osThreadDetach are not supported.

    Mutex
       osMutexPrioInherit is supported by default and is not configurable,
       we cannot select/unselect this attribute.
       osMutexRecursive is also supported by default. If this attribute is
       not set, an error is thrown when the same thread tries to acquire
       it the second time.
       osMutexRobust is not supported in Zephyr.

    Return values not supported in the Zephyr implementation
    ********************************************************

    osSemaphoreDelete
       osErrorResource (the semaphore specified by parameter
       semaphore_id is in an invalid semaphore state) is not supported.

    osMutexDelete
       osErrorResource (mutex specified by parameter mutex_id
       is in an invalid mutex state) is not supported.

    osTimerDelete
       osErrorResource (the timer specified by parameter timer_id
       is in an invalid timer state) is not supported.

    osMessageQueueReset
       osErrorResource (the message queue specified by
       parameter msgq_id is in an invalid message queue state)
       is not supported.

    osMessageQueueDelete
       osErrorResource (the message queue specified by
       parameter msgq_id is in an invalid message queue state)
       is not supported.

    osMemoryPoolFree
       osErrorResource (the memory pool specified by
       parameter mp_id is in an invalid memory pool state) is
       not supported.

    osMemoryPoolDelete
       osErrorResource (the memory pool specified by
       parameter mp_id is in an invalid memory pool state) is
       not supported.

    osEventFlagsSet, osEventFlagsClear
       osFlagsErrorUnknown (Unspecified error)
       and osFlagsErrorResource (Event flags object specified by
       parameter ef_id is not ready to be used) are not supported.

    osEventFlagsDelete
       osErrorParameter (the value of the parameter ef_id is
       incorrect) is not supported.

    osThreadFlagsSet
       osFlagsErrorUnknown (Unspecified error) and
       osFlagsErrorResource (Thread specified by parameter
       thread_id is not active to receive flags) are not supported.

    osThreadFlagsClear
       osFlagsErrorResource (Running thread is not active to
       receive flags) is not supported.

rgundi replied:
Done.

rgundi commented:
@pfalcon @galak @nashif Any final note on the naming stuff? I personally prefer "lib/ports_api/cmsis_rtos_v2" for C files and "include/ports_api/cmsis_rtos_v2" for header files.

rgundi commented:
@andyross The shippable failure is due to the multi-queue scheduler supporting a max of only 32 priority levels which is less than the total priority levels supported for cmsis_rtos_v2 (which is 56). Could you please look into this and see if that can be doubled to 64?

dbkinder reviewed (approved):
+1 for docs

nashif commented:
> Oh, and let's not forget that POSIX is also "compatibility/portability API" in a sense, so any ideas re: CMSIS should also be tested for applicability to existing POSIX stuff.

I am not sure what you mean here. Those are 2 different interfaces that sit on top of the native zephyr kernel APIs, stacking them on top of each other is in scope here.

nashif commented:
> @pfalcon @galak @nashif Any final note on the naming stuff? I personally prefer "lib/ports_api/cmsis_rtos_v2" for C files and "include/ports_api/cmsis_rtos_v2" for header files.

Maybe...

- interface/adaption
  - posix
  - cmsis_v1
  - cmsis_v2

Please address any remaining issues beside the naming of the directory, we can eventually get more input in the TSC.

rgundi commented:
@nashif addressed all review comments so far. Waiting for @andyross review especially regarding shippable failure for qemu_cortex_m3 which is due to the multi-queue scheduler.

rgundi committed:
tests/cmsis_rtos_v2: Use CONFIG_SCHED_SCALABLE

Using multi-queue scheduler doesn't work for cmsis_rtos_v2 as it
supports a max of only 32 priority levels which is less than the
total priority levels required by cmsis_rtos_v2 (which is 56).
Hence explicitly asking the tests to use the scalable scheduler.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi commented:
@nashif... Addressed the comments. The shippable is passing.

nashif reviewed (approved):
@rgundi I do not see a single sample using this or the v1 APIs, we do need samples, for example implementing simple synchronization using this API. This was part of the acceptance criteria.

nashif reviewed (changes requested):
@rgundi I do not see a single sample using this or the v1 APIs, we do need samples, for example implementing simple synchronization using this API. This was part of the acceptance criteria.

rgundi commented:
@nashif It escaped my notice but your request does make sense. Can that be done in a different PR? That PR can address both v1 and v2.

dbkinder reviewed (approved):
doc change comments from #11648 included here, so LGTM

spoorthik commented on lib/cmsis_rtos_v2/kernel.c:
```diff
@@ -0,0 +1,138 @@
+ */
+int32_t osKernelLock(void)
+{
+	int temp = _current->base.sched_locked;
```
This API cannot be called from ISR context. I think we need a check something like this here:

```
if (k_is_in_isr()) {
        return osErrorISR;
    }
```

rgundi replied:
addressed. Thanks for pointing out.

spoorthik reviewed (commented):
osKernelLock(), osKernelUnlock(), osKernelRestoreLock() APIs cannot be called from ISR context. So necessary check needs to be done. Please have a look.

nashif reviewed (approved):
naming of folder should not block this any further, we can rename where this is located along side v1 and posix when we have a better location and we get to the point of header directory structure cleanup

rgundi commented:
If there are any checkpatch errors, they are due to the external file cmsis_os2.h. Please ignore them.
